File: ValidParentheses/valid_parentheses.cpp
#include<iostream>
#include<stack>
#include<string>
using namespace std;

bool isValid(string s){
    stack<char> st;

    for (char &c : s){
        if (c == '(' || c == '{' || c == '['){
            st.push(c);
        }
        else{
            if (
                st.empty() ||
                (c == ')' && st.top() != '(') ||
                (c == ']' && st.top() != '[') ||
                (c == '}' && st.top() != '{')                 
            )
            {
                return false;
            }
            else{
                st.pop();
            }
        }
    }
    return st.empty();
}

int main(){
    string s = "()[]{}";
    cout << isValid(s) << endl;
    return 0;
}

File: ValidParentheses/valid_parentheses.py
def isValid(s: str) -> bool:
    stack = []
    for i in s:
        if i in "([{":
            stack.append(i)
        else:
            if not stack \
                or (i == ')' and stack[-1] != '(') \
                or (i == ']' and stack[-1] != '[') \
                or (i == '}' and stack[-1] != '{'):
                return False
            stack.pop()
        
    return not stack


if __name__ == "__main__":
    print(isValid("()"))        # True
    print(isValid("()[]{}"))    # True
    print(isValid("(]"))        # False
    print(isValid("([)]"))      # False
    print(isValid("{[]}"))      # True
    print(isValid("]"))         # False
    print(isValid("(("))        # False